compiler/graindoc/docblock.re

```diff
@@ -89,61 +103,60 @@ and provided = {
-exception MissingUnlabeledParamType({idx: int});
-exception MissingReturnType;
-exception AttributeAppearsMultipleTimes({attr: string});
-exception
-  InvalidAttribute({
-    name: string,
-    attr: string,
-  });
+type error =
+  | MissingFlag({
+      flag: string,
+      attr: string,
+    })
+  | MissingLabeledParamType({name: string})
+  | MissingUnlabeledParamType({idx: int})
+  | MissingReturnType
+  | AttributeAppearsMultipleTimes({attr: string})
+  | InvalidAttribute({
+      name: string,
+      attr: string,
+    });
+
+exception Error(Location.t, error);
+
+let report_error = (ppf, err) => {
+  switch (err) {
+  | MissingFlag({flag, attr}) =>
+    Format.fprintf(
+      ppf,
+      "Must provide %s when generating docs with `%s` attribute.",
+      flag,
+      attr,
+    )
+  | MissingLabeledParamType({name}) =>
+    Format.fprintf(
+      ppf,
+      "Unable to find a matching function parameter for %s. Make sure a parameter exists with this label or use `@param <param_index> %s` for unlabeled parameters.",
+      name,
+      name,
+    )
+  | MissingUnlabeledParamType({idx}) =>
+    Format.fprintf(
+      ppf,
+      "Unable to find a type for parameter at index %d. Make sure a parameter exists at this index in the parameter list.",
+      idx,
+    )
+  | MissingReturnType =>
+    Format.fprintf(ppf, "Unable to find a return type. Please file an issue!")
+  | AttributeAppearsMultipleTimes({attr}) =>
+    Format.fprintf(ppf, "Attribute @%s is only allowed to appear once.", attr)
+  | InvalidAttribute({name, attr}) =>
+    Format.fprintf(ppf, "Invalid attribute @%s on %s", attr, name)
+  };
+};
 
 let () =
-  Printexc.register_printer(exn => {
-    switch (exn) {
-    | MissingFlag({flag, attr}) =>
-      let msg =
-        Printf.sprintf(
-          "Must provide %s when generating docs with `%s` attribute.",
-          flag,
-          attr,
-        );
-      Some(msg);
-    | MissingLabeledParamType({name}) =>
-      let msg =
-        Printf.sprintf(
-          "Unable to find a matching function parameter for %s. Make sure a parameter exists with this label or use `@param <param_index> %s` for unlabeled parameters.",
-          name,
-          name,
-        );
-      Some(msg);
-    | MissingUnlabeledParamType({idx}) =>
-      let msg =
-        Printf.sprintf(
-          "Unable to find a type for parameter at index %d. Make sure a parameter exists at this index in the parameter list.",
-          idx,
-        );
-      Some(msg);
-    | MissingReturnType =>
-      let msg = "Unable to find a return type. Please file an issue!";
-      Some(msg);
-    | AttributeAppearsMultipleTimes({attr}) =>
-      let msg =
-        Printf.sprintf("Attribute @%s is only allowed to appear once.", attr);
-      Some(msg);
-    | InvalidAttribute({name, attr}) =>
-      let msg = Printf.sprintf("Invalid attribute @%s on %s", attr, name);
-      Some(msg);
-    | _ => None
-    }
-  });
+  Location.register_error_of_exn(
+    fun
```

Comment:
Nit I thought we generally avoid using `fun`

Reply:
No fun allowed! This is how we do all of the `register_error_of_exn`s so I just copied it, but I'll change it if you want.